scrapy_poet/_request_fingerprinter.py

```diff
@@ -0,0 +1,102 @@
+try:
+    from scrapy.utils.request import RequestFingerprinter  # NOQA
+except ImportError:
+    from typing import TYPE_CHECKING
+
+    if not TYPE_CHECKING:
+        ScrapyPoetRequestFingerprinter = None
+else:
+    import hashlib
+    import json
+    from functools import cached_property
+    from typing import Callable, Dict, List, Optional, get_args, get_origin
+    from weakref import WeakKeyDictionary
+
+    from scrapy import Request
+    from scrapy.crawler import Crawler
+    from scrapy.settings.default_settings import REQUEST_FINGERPRINTER_CLASS
+    from scrapy.utils.misc import create_instance, load_object
+
+    from scrapy_poet import InjectionMiddleware
+    from scrapy_poet.injection import get_callback
+
+    def _serialize_dep(cls):
+        try:
+            from typing import Annotated
+        except ImportError:
+            pass
+        else:
+            if get_origin(cls) is Annotated:
+                annotated, *annotations = get_args(cls)
+                return f"{_serialize_dep(annotated)}{repr(annotations)}"
```

Comment:
@kmike Is this `repr` OK?

Reply:
It should work fine with the current features, but there are caveats to keep in mind - the repr of annotated values shouldn't contain ids. Python docs mention addresses commonly being a part of repr, see https://docs.python.org/3/library/functions.html#repr. For example, if we define an annotation like this (likely we won't be adding geolocation support this way though), it'd break:

```py
class Geolocation:
    def __init__(self, geo: str):
        self.geo = geo

@attrs.define
class MyPage(ItemPage):
    resp: Annotated[HttpResponse, [Geolocation("UK")]]
```

To fix it, developer should override `__repr__` of the Geolocation class; it looks non-obvious than `__repr__` changes the fingerprinting behavior. 

I'm not sure what's the best fix here. One option is to require annotations to be json-serializable, but that's invasive, as even Enum values are not serializable. 

We can also solve it with a piece of documentation, to mention that repr is used for fingerprints. 

It may also be an option to add support for some other special method to be used to serialize annotations, which the classes can define, and fallback to repr if it's not available. It may also be a singledispatch hook, not a method.

I'd probably go with docs approach for now, because the issue doesn't seem to affect any of the current or planned features, but I'm not 100% sure.

Reply:
[Docs for now it is](https://github.com/scrapinghub/scrapy-poet/pull/172/commits/596431b1c271c13144b1fd45546261b53bba5cb3).